---
title: "Expression_Analysis_Markdown"
author: "Devin Molnau and Katerina Holan"
date: "April 25, 2017"
output: html_document
---
#Expression Analysis Markdown

Figure 1 is a multidimensional scaling plot that represents the distance between all samples of Asparagus officinalis as the root-mean-square deviation. This was done to the top 250 most heterogeneously expressed genes in a pairwise comparison.

Figure 2 is comprised of three pairwise analyses gene expression among female, male and supermale garden asparagus (Asparagus officinalis) genes across all lines (8A,8B, 10, 9). Average log fold change (FC) vs average log of counts per million mapped reads, CPM

Figure 3 is a Venn diagram of the differentially expressed genes found in the pairwise analyses.

Figure 4 is a heatmap clustering of the differentially expressed genes in garden asparagus in spear tip tissue.

###import bioconductor and required libraries
```{r}
source("http://bioconductor.org/biocLite.R")
biocLite()
biocLite("edgeR") 
library(edgeR)
```

###Import counts data
```{r}

#read in the gene counts from paper
counts_data<-read.delim("Asparagus.RSEM_genecounts.txt", check.names=FALSE, stringsAsFactors = FALSE)
```

###convert gene count data to DGEList
```{r}
cd<-DGEList(counts=counts_data)
```
###Filtering

Filters transcripts from read counts data. The transcripts that are kept must belong in at least three libraries. Each library contains more than one count per million mapped reads.
```{r}
keep <- rowSums(cpm(cd)>1) >= 3 #filters out lowly expressed genes
cd <- cd[keep, , keep.lib.sizes=FALSE]
```
###Normalized data

Normalizing data takes into the effect of the library size on the samples. Scaling factors are found for the each of the libraries that minimize the log-fold changes between the samples for most genes. 
The calcNormFactor() uses trimmed mean of M values to determine the scaling factors. These scaling factors in turn can be used to find the effective library size. 
```{r}
cd<-calcNormFactors(cd) #normalizes using TMM
```

###Figure 1: Multidimentional scaling 

Figure 1 is a multidimensional scaling plot of all lines
```{r}
colors <- c("black","black","black","darkgreen","darkgreen","darkgreen","blue","blue","blue", "red", "red") #colors for the lines
#MDS plot without legend, but the names indicate line and sex
plotMDS(cd, col=colors, top = 250, gene.selection = "pairwise", main="MDS plot of various lines and sex of Asparagus officinalis") 
```
###Design Matrix

Additive model of sex and line for fitting. This accounts for differences due to line rather than sex in pairwise analysis.
```{r}
#design matrix of sex and line
rm(design)
line<-factor(c(88,88,88,89,89,89,103,103,103,9,9))
sex<-factor(c("XX","XY","YY","XX","XY","YY","XX","XY","YY","XX","XY"))
#sex<-factor(c("female", "male","supermale","female", "male","supermale","female", "male","supermale","female", "male"))
data.frame(Sample=colnames(cd),line,sex)
design <- model.matrix(~line+sex)
rownames(design) <- colnames(cd)
design
```
###Estimating Dispersion

edgeR packaged uses Cox-Reid profile-adjusted likelihood (CR) method in estimating dispersions when there are multiple factors (sex and line, in this case).
```{r}
cd <- estimateGLMCommonDisp(cd, design) #estimates common dispersion
cd <- estimateGLMTrendedDisp(cd, design)#estimates trended dispersion
cd <- estimateGLMTagwiseDisp(cd, design)#estimates tagwise dispersion
```

###Figure 2: Testing for Differential Expression

glmFit() takes gene counts, dispersions, and the design matrix as an input and produces a DGEGLM object as an output. This fits the negative binomial GLM for each gene id.
```{r}
fit <- glmFit(cd, design)
```

Pairwise comparison between Female and Supermale groups. Note that the coef=6 in glmLRT refers to the 6th column on the design matrix, thus comparing Females from the first column to supermales in the 6th column.
Adjust P value using false discovery rate (FDR) method and then determine how many genes are significantly differentially expressed between female and supermale asparagus. 
Pairwise comparison is plotted using plotMD().
```{r}
lrt_FSupM <- glmLRT(fit, coef=6)  #Female VS Supermale
summary(de_FemvsSupMal <- decideTestsDGE(lrt_FSupM, adjust.method="fdr"))
FDR_FSupM <- p.adjust(lrt_FSupM$table$PValue, method="fdr") 
sum(FDR_FSupM < 0.05) #408  differentially expressed genes (462 if not normalized)
status<-rep("not significant",length(FDR_FSupM))
status[FDR_FSupM<0.05 & lrt_FSupM$table$logFC>0]<-"up"
status[FDR_FSupM<0.05 & lrt_FSupM$table$logFC<0]<-"down"
plotMD(lrt_FSupM, main="SuperMale VS Female",hl.col=c("red","blue"),status = status, bg.col="grey")
```

Pairwise comparison between Female and Male groups. Note that the coef=5 in glmLRT refers to the 5th column on the design matrix, thus comparing Females from the first column to Males in the 5th column.
Adjust P value using false discovery rate (FDR) method and then determine how many genes are significantly differentially expressed between female and male asparagus. 
Pairwise comparison is plotted using plotMD().

```{r}
lrt_FM <- glmLRT(fit, coef=5) #female VS male
#lrt_FM <- glmLRT(fit, contrast=c(-1,0,0,0,1,0))
summary(de_FemvsMal <- decideTestsDGE(lrt_FM, adjust.method="fdr"))
FDR_FMale<-p.adjust(lrt_FM$table$PValue,method="fdr") #adjust Pvalue
sum(FDR_FMale<0.05) #221 significant differentially expressed genes
status<-rep("not significant",length(FDR_FMale))
status[FDR_FMale<0.05 & lrt_FM$table$logFC>0]<-"up"
status[FDR_FMale<0.05 & lrt_FM$table$logFC<0]<-"down"
plotMD(lrt_FM, main="Male VS Female",hl.col=c("red","blue"),status = status, bg.col="grey")
```
Pairwise comparison between Male and Supermale groups. Note that the fifth and sixth columns are Male and Supermale respectively and must be referred to by their position in the design matrix.
Adjust P value using false discovery rate (FDR) method and then determine how many genes are significantly differentially expressed between male and supermale asparagus. 
Pairwise comparison is plotted using plotMD().
```{r}
lrt_MSupM <- glmLRT(fit, contrast=c(0,0,0,0,-1,1)) #Male VS Supermale
summary(de_MalvsSupMal <- decideTestsDGE(lrt_MSupM, adjust.method="fdr"))
FDR_MSupMale<-p.adjust(lrt_MSupM$table$PValue,method="fdr")
sum(FDR_MSupMale<0.05) #27
status<-rep("not significant",length(FDR_MSupMale))
status[FDR_MSupMale<0.05 & lrt_MSupM$table$logFC>0]<-"up"
status[FDR_MSupMale<0.05 & lrt_MSupM$table$logFC<0]<-"down"
plotMD(lrt_MSupM, main="Supermale VS Male",hl.col=c("red","blue"),status = status, bg.col="grey")
```
###Figure 3: Venn Diagram

Making vectors of the genes that were signficantly differentially expressed in each pairwise comparison.

The following returns a vector called names_of_FSupM_hit that contains all the gene_id names that were significantly differentially expressed between female and supermale asparagus.
```{r}
##Get the names of all of the genes that have a Pvalue<0.05 for FDR_FSupM
total_FDR_FSupM<-length(FDR_FSupM)
total_FDR_FSupM
names_of_FSupM_hit<-c()
counter<-0
for (i in 1:total_FDR_FSupM){
  if (FDR_FSupM[i]<0.05 ){
    counter<-counter+1
    names_of_FSupM_hit[counter]=rownames(lrt_FSupM)[i]
  }
}
names_of_FSupM_hit
num_FSupM_hit<-length(names_of_FSupM_hit)
num_FSupM_hit #408 
```

The following returns a vector called names_of_FMale_hit that contains all the gene_id names that were significantly differentially expressed between female and male asparagus.
```{r}
##Get the names of all of the genes that have a Pvalue<0.05 for FDR_FMale
total_FDR_FMale<-length(FDR_FMale) 
total_FDR_FMale
names_of_FMale_hit<-c() #empty vector
counter<-0
for (i in 1:total_FDR_FMale){
  if (FDR_FMale[i]<0.05 ){
    counter<-counter+1
    names_of_FMale_hit[counter]=rownames(lrt_FM)[i] #adds the gene_id to empty vector if significantly differently expressed
  }
}
names_of_FMale_hit
num_FMale_hit<-length(names_of_FMale_hit)
num_FMale_hit #221 DF genes in Female/Male pairwise comparison
```

The following returns a vector called names_of_MSupMale_hit that contains all the gene_id names that were significantly differentially expressed between male and supermale asparagus.
```{r}
##Get the names of all of the genes that have a Pvalue<0.05 for FDR_FMale
total_FDR_MSupMale<-length(FDR_MSupMale)
total_FDR_MSupMale
names_of_MSupMale_hit<-c()
counter<-0
for (i in 1:total_FDR_MSupMale){
  if (FDR_MSupMale[i]<0.05 ){
    counter<-counter+1
    names_of_MSupMale_hit[counter]=rownames(lrt_MSupM)[i]
  }
}
names_of_MSupMale_hit
num_MSupMale_hit<-length(names_of_MSupMale_hit)
num_MSupMale_hit #27
```
Comparison of differentially expressed genes in Female/Supermale and Male/Supermale pairwise comparison. 
```{r}
###Checks the number of hits between FSupM and MSupMale
FSupM_MSupM<-names_of_FSupM_hit %in% names_of_MSupMale_hit #compares the two vectors of gene names
sum(FSupM_MSupM) #23 common DF genes shared
num_FSupM_MSupM<-sum(FSupM_MSupM) #makes variable for venn diagram
```
Comparison of differentially expressed genes in Female/Male and Female/Supermale
pairwise comparisons.
```{r}
FMale_FSupM<-names_of_FMale_hit %in% names_of_FSupM_hit
sum(FMale_FSupM)#150 differential genes shared between Female/Male and Female/Supermale
num_FMale_FSupM<-sum(FMale_FSupM) #makes variable for venn diagram
```
Comparison of differentially expressed genes in Female/Male and Male/Supermale pairwise comparisons
```{r}
FMale_MSupMale<-names_of_FMale_hit %in% names_of_MSupMale_hit
sum(FMale_MSupMale)#2 DF genes 
num_FMale_MSupMale<-sum(FMale_MSupMale) #makes variable for venn diagram
```
Comparison of significantly differentially expressed gene in Female/Male, Male/Supermale, Female/Supermale pairwise comparisons.
```{r}
FSupM_MSupM_FMale<-names_of_FMale_hit %in% names_of_FSupM_hit %in% names_of_MSupMale_hit
sum(FSupM_MSupM_FMale)
num_FSupM_MSupM_FMale<-sum(FSupM_MSupM_FMale) #0 DF genes shared between all three groups
```

Generating the VennDiagram
```{r}
###Venn Diagram###
install.packages('VennDiagram')
library(VennDiagram)
grid.newpage()
draw.triple.venn(area1 = num_FMale_hit, area2 = num_FSupM_hit, 
                 area3 = num_MSupMale_hit, n12 = num_FMale_FSupM,
                 n23 = num_FSupM_MSupM, n13 = num_FMale_MSupMale, 
                 n123 = num_FSupM_MSupM_FMale, 
                 category = c("Male vs Female", "Supermale vs female",
                              "Supermale vs male"),lty = "blank", 
                 fill = c("skyblue", "pink1", "mediumorchid"),
                 cex = 1, cat.cex = 1.5,
                 cat.just=list(c(-0.2,4) , c(1,4) , c(0.5,-4))) 
                #dimensions courtesy of Kat

```


Alternative code for generating Venn diagram
```{r}
#getting genenames for the DE genes in Female vs supermale comparison 
detags_FSupM <- rownames(cd)[as.logical(de_FemvsSupMal)]
#female vs male
detags_FM <- rownames(cd)[as.logical(de_FemvsMal)]
#male vs supermale
detags_MSupM <- rownames(cd)[as.logical(de_MalvsSupMal)]

#Getting numbers for each section of the Venn diagram
FM <- length(detags_FM)
FSupM <- length(detags_FSupM)
MSupM <- length(detags_MSupM)
FM_FSupM <- sum(detags_FM %in% detags_FSupM)
FM_MSupM <- sum(detags_FM %in% detags_MSupM)
FSupM_MSupM <- sum(detags_FSupM %in% detags_MSupM)
FM_FSupM_MSupM <- sum(detags_FM %in% detags_FSupM %in% detags_MSupM)

#generating Venn diagram
#install.packages('VennDiagram')
library(VennDiagram)
grid.newpage()
draw.triple.venn(area1 = FM, area2 = FSupM, area3 = MSupM, 
                 n12 = FM_FSupM, n23 = FSupM_MSupM, n13 = FM_MSupM, 
                 n123 = FM_FSupM_MSupM, 
                 category = c("Male vs Female", "Supermale vs female", "Supermale vs male"),
                 lty = "blank", 
                 fill = c("skyblue", "pink1", "mediumorchid"),
                 cex = 2, cat.cex = 2,
                 cat.just=list(c(-0.2,4) , c(1,4) , c(0.5,-4)))
```

###DE Genes Breakdown
Breakdown of gene names from each comparison from our analysis and from the paper's analysis, as well as a comparison of the DE genes from each comparison.
```{r}
#DE genes for each comparison
detags_FSupM <- rownames(cd)[as.logical(de_FemvsSupMal)]
detags_FM <- rownames(cd)[as.logical(de_FemvsMal)]
detags_MSupM <- rownames(cd)[as.logical(de_MalvsSupMal)]

###total number/names of unique DEGs in our analysis
detags <- c(detags_FM, detags_FSupM, detags_MSupM)
detags_df <- as.data.frame(detags)
uniq_detags <- unique(detags_df)
colnames(uniq_detags) <- "gene_id"

###DEGs from paper
paper_fsupm <- read.delim("paper_fsupm.txt")
paper_fm <- read.delim("paper_fm.txt")
paper_msupm <- read.delim("paper_msupm.txt")
paper_detags <- rbind(paper_fsupm, paper_fm, paper_msupm)
uniq_paper_detags <- unique(paper_detags)
in_both <- sum(sort(uniq_detags[,]) %in% sort(uniq_paper_detags[,]))
###475 common between our analysis and theirs

###Breakdown of genes that matched between our analysis and the paper's
compare_FM <- sum(detags_FM %in% paper_fm$gene_id)
compare_FSupM <- sum(detags_FSupM %in% paper_fsupm$gene_id)
compare_MSupM <- sum(detags_MSupM %in% paper_msupm$gene_id)

```


###Figure 4:  Heatmap
workflow from https://github.com/alexharkess/PBIO_6550/wiki/Lab-8-:-Data-munging-and-visualization

Tutorial heatmap using 9 out of 11 lines (missing 8A male and 10 supermale).  TMM normalized FPKM values for the 570 DE genes determined in the paper were used to generate the heatmap.
```{r}
library(ggplot2)
#install.packages(c("reshape", "gplots"))
library(reshape)
library(gplots)

FPKM_matrix <- read.table("TMM_normalized_FPKM_matrix.txt", header = T, row.names = 1)
FPKM_matrix_melt <- melt(FPKM_matrix)
plot <- ggplot(FPKM_matrix_melt, aes(x = variable, y = value)) + geom_boxplot() +
  geom_jitter()
plot
plot + ylim(0,100)

FPKM_matrix <- as.matrix(FPKM_matrix)

my_palette <- colorRampPalette(c("turquoise4","white","maroon2"))(n = 1000)
heatmap.2(FPKM_matrix, dendrogram="both", trace="none", scale="row", density.info="none", col=my_palette,
          srtCol=45, cexCol = 1)

```

Since our assembly analysis only looked at 2 samples, we made a heatmap of just the counts data, instead of the FPKM values.
Pulling out only genes we found to be DE from the counts data
```{r}
library(ggplot2)
#install.packages(c("reshape", "gplots"))
library(reshape)
library(gplots)

cd_matrix <- as.matrix(cd)
cd_df <- as.data.frame(cd_matrix)
genenames <- uniq_detags$gene_id
library(data.table)
setDT(cd_df, keep.rownames = TRUE)[]
#DEG_counts <- cd_df[cd_df$rn %in% genenames,]
#write.csv(DEG_counts, file = "DEG_counts.csv")
heatmap_counts <- read.csv("DEG_counts.csv", header = T, row.names = "rn")

```

Importing the csv switched the row names column with the row numbers.  This removes the extra column of numbers and changes the heatmap to a matrix which is needed to input into the heatmap function.
```{r}
###introduces column with row numbers from original counts data frame
heatmap_counts <- heatmap_counts[,2:12]
heatmap_counts_matrix <- as.matrix(heatmap_counts)

```

Setting color palette for the heatmap, creating heatmap.
```{r}
my_palette <- colorRampPalette(c("turquoise4","white","maroon2"))(n = 1000)

heatmap.2(heatmap_counts_matrix, dendrogram="both", trace="none", scale="row", density.info="none", col=my_palette, srtCol=45, cexCol = 1)

```


###Creating heatmap from samples that went through transcriptome pipeline
read in the 2 files with FPKM values
```{r}
#read in files with FPKM.  from alignment counting folder
samp1 <- read.delim("SRR1639681_counting.genes.results", stringsAsFactors = FALSE)
samp2 <- read.delim("SRR1642915_counting.genes.results", stringsAsFactors = FALSE)
 
```

Join fpkm columns from each sample together into one data frame
```{r}
#rename fpkm columns with sample name

colnames(samp1)[6] <- "FPKM_1"
colnames(samp2)[6] <- "FPKM_2"

joined_fpkm <- cbind(samp1$gene_id, samp1$FPKM_1, samp2$FPKM_2)
joined_fpkm <- as.data.frame(joined_fpkm)

```

Rename rows with the gene_id names and label columns with sample numbers
```{r}
fpkm_rownames <- data.frame(joined_fpkm[,-1], row.names=joined_fpkm[,1])
fpkm_colnames <- fpkm_rownames
colnames(fpkm_colnames)[1] <- "FPKM_1"
colnames(fpkm_colnames)[2] <- "FPKM_2"

```

gene id names are different format in uniq_detags and fpkm data frame so need to change uniq_detags to match those in fpkm (or vice-versa).  This turns out to be fairly easy with a simple find and replace function gsub()
```{r}
uniq_detags_vector <- as.vector(uniq_detags$gene_id)
uniq_detags_aspof <- gsub("comp", "Aspof_comp", uniq_detags_vector)
```

subset fpkm values by only those determined to be DEG
```{r}
deg_fpkm <- subset(fpkm_colnames, rownames(fpkm_colnames) %in% uniq_detags_aspof)
#make sure FPKM values are numeric, NOT factors
deg_fpkm$FPKM_1 <- as.numeric(as.character(deg_fpkm$FPKM_1))
deg_fpkm$FPKM_2 <- as.numeric(as.character(deg_fpkm$FPKM_2))

```

make into matrix and generate heatmap
```{r}
#since FPKM values are strongly right-tailed with few very large numbers, a log2 transformation is done to better visualize the data.  1 is added the each value to prevent the 0s from being changed to infinity values
deg_fpkm_log <- log2(deg_fpkm[,] + 1)
#use summary to decide where you want to have your breaks.
summary(deg_fpkm_log)
#breaks sets how many colors in your gradient and what values get assigned what colors.  Max is 10.567
breaks=seq(0, 11, by=0.1)

deg_fpkm_matrix <- as.matrix(deg_fpkm_log)

my_col <- colorpanel(n=length(breaks)-1,low="gray75",high="black")

heatmap.2(deg_fpkm_matrix, breaks = breaks, dendrogram = "none", trace="none", density.info="none", col=my_col, srtCol=45, cexCol = 1)

```




